### PR TITLE
feat(photomath): `unlock-plus` patch

### LINKED ISSUE #1461: Unlock Photomath plus features
### Application

[Photomath](https://play.google.com/store/apps/details?id=com.microblink.photomath) 

### Issue

Unlock Photomath plus features

### Patch

The patch should unlock in-depth explanations and animated tutorials.

### Motivation

Photomath plus provides in-depth explanations when solving math problems. Would be extremely useful during study.

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.

### LINKED ISSUE #1461: Unlock Photomath plus features
### Application

[Photomath](https://play.google.com/store/apps/details?id=com.microblink.photomath) 

### Issue

Unlock Photomath plus features

### Patch

The patch should unlock in-depth explanations and animated tutorials.

### Motivation

Photomath plus provides in-depth explanations when solving math problems. Would be extremely useful during study.

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.

### PR DESCRIPTION
**Pull request by nik2143**
Unlock features of photomath plus
Closes ReVanced/revanced-patches-template#1461

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -1,2 +1,2 @@
 kotlin.code.style = official
-version = 2.162.0-dev.3
+version = 2.162.0
```
Comment (oSumAtrIX) [inline gradle.properties:2]: This file is autogenerated, do not commit it.

File: src/main/kotlin/app/revanced/patches/photomath/detection/signature/fingerprints/MainOnCreateingerprint.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package app.revanced.patches.photomath.detection.signature.fingerprints
+
+import app.revanced.patcher.extensions.or
+import app.revanced.patcher.fingerprint.method.impl.MethodFingerprint
+import org.jf.dexlib2.AccessFlags
+
+object MainOnCreateingerprint : MethodFingerprint(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/detection/signature/fingerprints/MainOnCreateingerprint.kt:7]:
```suggestion
object MainOnCreateFingerprint : MethodFingerprint(
```

File: src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.photomath.detection.signature.annotations.DisableSignatureDetectionCompatibilty
+import app.revanced.patches.photomath.detection.signature.fingerprints.MainOnCreateingerprint
+import org.jf.dexlib2.Opcode
+import org.jf.dexlib2.builder.instruction.BuilderInstruction11n
+import org.jf.dexlib2.iface.instruction.OneRegisterInstruction
+
+@Patch
+@Name("disable-signature-detection")
+@Description("Disable the signature detection.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt:22]:
```suggestion
@Description("Disables detection of incorrect signature.")
```

File: src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+@DisableSignatureDetectionCompatibilty
+@Version("0.0.1")
+class SignatureDetectionPatch : BytecodePatch(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt:25]: Please format the code.

File: src/main/kotlin/app/revanced/patches/photomath/misc/unlockplus/patch/UnlockPlusPatch.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import app.revanced.patcher.patch.annotations.DependsOn
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.photomath.detection.signature.patch.SignatureDetectionPatch
+import app.revanced.patches.photomath.misc.unlockplus.annotations.UnlockPlusCompatibilty
+import app.revanced.patches.photomath.misc.unlockplus.fingerprints.IsPlusUnlockedFingerprint
+
+@Patch
+@Name("unlock-plus")
+@DependsOn([SignatureDetectionPatch::class])
+@Description("Unlock plus features.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/misc/unlockplus/patch/UnlockPlusPatch.kt:21]:
```suggestion
@Description("Unlocks plus features.")
```

File: src/main/kotlin/app/revanced/patches/photomath/misc/unlockplus/patch/UnlockPlusPatch.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+@UnlockPlusCompatibilty
+@Version("0.0.1")
+class UnlockPlusPatch : BytecodePatch(
+    listOf(
+        IsPlusUnlockedFingerprint
+    )
+){
+    override fun execute(context: BytecodeContext): PatchResult {
+        val plusUnlockedMethod = IsPlusUnlockedFingerprint.result?.mutableMethod
+            ?: return PatchResultError("Can't find isPlusUnlocked method")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/misc/unlockplus/patch/UnlockPlusPatch.kt:31]:
```suggestion
            ?: return IsPlusUnlockedFingerprint.toErrorResult()
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
Seems like you [overwrote](https://github.com/revanced/revanced-patches/compare/e264941bbfc4efe7d55addb5ff979ff643ef3df0..983c418c58877bd73f83170c11ef54e0cac33289) my [changes](https://github.com/revanced/revanced-patches/commit/983c418c58877bd73f83170c11ef54e0cac33289) with [your force push](https://github.com/revanced/revanced-patches/commit/983c418c58877bd73f83170c11ef54e0cac33289). 
Please revert your force-push via the reflog and merge or rebase your review suggestions on top of my changes, as they contain various fixes, your force-push has overwritten.

File: src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                Opcode.INVOKE_STATIC,
+                Opcode.INVOKE_STATIC,
+                Opcode.MOVE_RESULT_OBJECT,
+                Opcode.INVOKE_VIRTUAL,
+                Opcode.MOVE_RESULT_OBJECT,
+                Opcode.INVOKE_STATIC,
+            )
+        ) {}
+        val patternResult = checkSignatureFingerprint.also {
+            it.resolve(context,onCreateMethod!!,MainOnCreateFingerprint.result?.classDef!!)
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt:48]: Please format your code.

File: src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import app.revanced.patcher.patch.PatchResultError
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.photomath.detection.signature.annotations.DisableSignatureDetectionCompatibilty
+import app.revanced.patches.photomath.detection.signature.fingerprints.MainOnCreateFingerprint
+import org.jf.dexlib2.Opcode
+import org.jf.dexlib2.builder.instruction.BuilderInstruction11n
+import org.jf.dexlib2.iface.instruction.OneRegisterInstruction
+
+@Patch
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt:20]: Is this meant to be a user-facing patch? It doesn't really serve any purpose in itself.

File: src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        val patternResult = checkSignatureFingerprint.also {
+            it.resolve(context,onCreateMethod!!,MainOnCreateFingerprint.result?.classDef!!)
+        }.result!!.scanResult.patternScanResult
+
+        with(onCreateMethod?.implementation) {
+            val signCheckResultInstr = this?.instructions?.get(patternResult?.endIndex!! + 1)
+            if (signCheckResultInstr?.opcode!=Opcode.MOVE_RESULT) return PatchResultError("Can't find result of signature check")
+            this?.replaceInstruction(signCheckResultInstr.location.index,
+                BuilderInstruction11n(Opcode.CONST_4,(signCheckResultInstr as OneRegisterInstruction).registerA,1))
+        }
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt:56]: There is quite a lot going on here, especially with nullability and null assertions, might want to look into reducing noise here.

**Comment by oSumAtrIX** [comment]
Pushed some refactoring changes, please test so we can merge.

**Review by oSumAtrIX — APPROVED** [review]

**Comment by nik2143** [comment]
Working, can be merged

File: src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    )
+) {
+    override fun execute(context: BytecodeContext): PatchResult {
+        val mainOnCreate = MainOnCreateFingerprint.result!!
+
+        val patternResult = CheckSignatureFingerprint.also {
+            it.resolve(context, mainOnCreate.method, mainOnCreate.classDef)
+        }.result!!.scanResult.patternScanResult!!
+
+        with(mainOnCreate.mutableMethod) {
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/photomath/detection/signature/patch/SignatureDetectionPatch.kt:32]:
```suggestion
        mainOnCreate.mutableMethod.apply {
```

**Comment by oSumAtrIX** [comment]
Thanks

**Pull request merged**